[UR][L0] Manage UMF pools through usm::pool_manager

kswiecicki opened the pull request:
PR from: https://github.com/oneapi-src/unified-runtime/pull/2495

kswiecicki commented:
@igchor, I've had an issue with obtaining the pool config `InitializeDisjointPoolConfig()` locally in the usm pool constructor and passing it to for disjoint pool creation. Some allocations in `test-e2e/USM/usm_pooling.cpp` weren't really pooled even though they should be and the test was failing. Copying the config from `DisjointPoolConfigInstance` global variable instead seems to work. 
Could you re-review this change please.

igchor commented:
> InitializeDisjointPoolConfig

Hm, that's strange. Were all UR tests passing? I've had some problems with this test and v2 adapter but it's because it expects certain logs, and UMF wasn't outputting them (this will be fixed once UR switches to using L0 loader API logging)

igchor commented:
> > Isn't UseUSMAllocator a global variable? Can't we just read it in contextCreate and decide whether to create ProxyPool or DisjointPool there? We would only have one defaultPool in the context this way.
> 
> I think I've fixed the final issue with E2E tests. Because of the change that creates either `ProxyPool` or `DisjointPool` at the startup but never both, the test ran with `SYCL_PI_LEVEL_ZERO_DISABLE_USM_ALLOCATOR=1` no longer needs to account for the allocations made by `DisjointPool` when querying the page size.

Okay, that makes sense. However, removing the CHECK-COUNT will break this test for L0 v2 adapter, because there we always query page size (it's done by L0 provider). Perhaps we could modify this to be:
```
CHECK-OPT: zeMemFree
CHECK-OPT: zeMemFree
CHECK-OPT: zeMemFree
CHECK-OPT: zeMemFree
CHECK-OPT: zeMemFree
CHECK-OPT: zeMemFree
CHECK-OPT: zeMemFree
CHECK-OPT: zeMemFree
```

kswiecicki commented:
> > > Isn't UseUSMAllocator a global variable? Can't we just read it in contextCreate and decide whether to create ProxyPool or DisjointPool there? We would only have one defaultPool in the context this way.
> > 
> > 
> > I think I've fixed the final issue with E2E tests. Because of the change that creates either `ProxyPool` or `DisjointPool` at the startup but never both, the test ran with `SYCL_PI_LEVEL_ZERO_DISABLE_USM_ALLOCATOR=1` no longer needs to account for the allocations made by `DisjointPool` when querying the page size.
> 
> Okay, that makes sense. However, removing the CHECK-COUNT will break this test for L0 v2 adapter, because there we always query page size (it's done by L0 provider). Perhaps we could modify this to be:
> 
> ```
> CHECK-OPT: zeMemFree
> CHECK-OPT: zeMemFree
> CHECK-OPT: zeMemFree
> CHECK-OPT: zeMemFree
> CHECK-OPT: zeMemFree
> CHECK-OPT: zeMemFree
> CHECK-OPT: zeMemFree
> CHECK-OPT: zeMemFree
> ```

That could be a lot of unaccounted for frees, but I don't see a better solution unless there's `CHECK-COUNT-8-OPT` available.

kswiecicki commented:
The tests failing on CI are unrelated to this PR and are currently failing on other PRs.
Precommit Linux/Windows jobs (eg. https://github.com/intel/llvm/actions/runs/13675062022/job/38235698849 and https://github.com/intel/llvm/actions/runs/13672691248/job/38227250897)
```
********************
Failed Tests (2):
  SYCL :: AOT/early_aot.cpp
  SYCL :: SpecConstants/2020/image_selection.cpp
```

```
********************
Failed Tests (1):
  SYCL :: USM/memops2d/copy2d_device_to_device.cpp
```

uditagarwal97 commented:
> @intel/llvm-gatekeepers could someone merge this please.

I don't know why `Check for private emails used in PRs` workflow didn't ran on this PR. @sarnex FYI.
@kswiecicki was this PR moved over from UR via a script? Also, please use intel/codeplay email to author this PR, currently, GH says:
```
This commit will be authored by 61325956+kswiecicki@users.noreply.github.com.
```

uditagarwal97 commented:
> > @intel/llvm-gatekeepers could someone merge this please.
> 
> I don't know why `Check for private emails used in PRs` workflow didn't ran on this PR. @sarnex FYI. @kswiecicki was this PR moved over from UR via a script? Also, please use intel/codeplay email to author this PR, currently, GH says:
> 
> ```
> This commit will be authored by 61325956+kswiecicki@users.noreply.github.com.
> ```

PR to fix the private email check workflow: https://github.com/intel/llvm/pull/17333

kswiecicki commented:
> > @intel/llvm-gatekeepers could someone merge this please.
> 
> I don't know why `Check for private emails used in PRs` workflow didn't ran on this PR. @sarnex FYI. @kswiecicki was this PR moved over from UR via a script? Also, please use intel/codeplay email to author this PR, currently, GH says:
> 
> ```
> This commit will be authored by 61325956+kswiecicki@users.noreply.github.com.
> ```

I've had a problem with the UR script, so I've moved the PR manually. It's weird that those commits seem to be authored from `61325956+kswiecicki@users.noreply.github.com`, since `git log` shows my intel email. I've tweaked some github profile options and rebased this PR on https://github.com/intel/llvm/pull/17333. The mail check workflow passes, so I guess it's fixed now.
